src/app/cluster-building-blocks/QuieterReporting.h

```diff
@@ -0,0 +1,194 @@
+ *
+ * The class can be configured via `SetPolicy` to have some/all of the common reasons
+ * for reporting (e.g. increment only, decrement only, change to/from zero).
+ *
+ * Changes of null to non-null or non-null to null are always considered dirty.
+ *
+ * It is possible to force mark the attribute as dirty (see `ForceMarkAsDirty`) such as
+ * for conditions like "When there is any increase or decrease in the estimated time
+ * remaining that was due to progressing insight of the server's control logic".
+ *
+ * Usage is simple:
+ *
+ * - Call `SetValue()` with the new value and current monotonic system timestamp
+ *   - There is an overload with a `SufficientChangePredicate` which will apply externally
+ *     provided checks between old/new value and time last marked dirty to allow for complex
+ *     rules like marking dirty less than once per second, etc.
+ * - *Maybe* call `ForceMarkAsDirty()` in some choice situations (e.g. know for sure it's
+ *    an important update, like at the edge of an operational state change).
+ * - Call `GetThenResetDirtyState()`. If it returns true, mark the attribute dirty, with the
+ *   method most suitable at the call site (e.g. `MatterReportingAttributeChangeCallback` call
+ *   or similar methods).
+ *
+ * @tparam T - the type of underlying numerical value that will be held by the class.
+ */
+template <typename T>
+class QuieterReportingAttribute
+{
+  public:
+    explicit QuieterReportingAttribute(const chip::app::DataModel::Nullable<T>& initialValue) : mValue(initialValue), mLastDirtyValue(initialValue) {}
+
+    using SufficientChangePredicate = std::function<bool(Timestamp /* previousDirtyTime */, Timestamp /* now */, const Nullable<T> & /* previousDirtyValue */, const Nullable<T> & /* newValue */)>;
+
+    /**
+     * @brief Factory to generate a functor for "attribute was last reported" at least `minimumDurationMillis` ago.
+     *
+     * @param minimumDurationMillis - number of millis needed since last marked as dirty before we mark dirty again.
+     * @return a functor usable for the `changedPredicate` arg of `SetValue()`
+     */
+    static SufficientChangePredicate GetPredicateForSufficientTimeSinceLastDirty(chip::System::Clock::Milliseconds64 minimumDurationMillis)
+    {
+      return [minimumDurationMillis](Timestamp previousDirtyTime, Timestamp now, const Nullable<T> &oldDirtyValue, const Nullable<T> &newValue) -> bool {
+        return (oldDirtyValue != newValue) && ((now - previousDirtyTime) >= minimumDurationMillis);
+      };
+    }
+
+    chip::app::DataModel::Nullable<T> value() const { return mValue; }
+    QuieterReportingPolicyFlags policy() const { return mPolicyFlags; }
+
+    void SetPolicy(QuieterReportingPolicyFlags policyFlags) { mPolicyFlags = policyFlags; }
+
+    /**
+     * When this returns true, attribute should be marked for reporting. Auto-resets to false after call.
+     */
+    bool GetThenResetDirtyState()
```

Comment:
Methods with compund statements (and/then/...) feel off. Why don't we have `IsDirty` and `ClearDirty`? I can replace `foo = bar.GetThenResetDirtyState` with `foo = bar.IsDirty(); bar.ClearDirty()`... slightly more verbose however significant less cognitive load reading.

Reply:
It also allows IsDirty in unit tests (or other code) without resetting, such as asserts or other logic.

Reply:
Resetting is a key part of the API and critical to avoid mistakes. Staying sorry because of a missed ClearDirty can lead to performance issues and regression of the while point of the thing.

Reply:
Renamed to `WasJustMarkedDirty` which is more accurate.

Reply:
`WasJust` seems to imply some temporality that the class does not have. We just have dirty and a timestamp.

This is a naming + concept issue. Leaning towards looking at atomics compareexchange/fetchset or a `Clear` returning the previous value.